File: implementations/android/client/app/src/main/java/com/theshuai/specus/android/PeerServiceBridge.java
package com.theshuai.specus.android;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.Objects;
import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Semaphore;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.logging.Logger;

final class PeerServiceBridge implements AutoCloseable {
    private static final int MAX_ACTIVE_CONNECTIONS = 64;
    private static final Logger LOG = Logger.getLogger(PeerServiceBridge.class.getName());
    private final String virtualIp;
    private final SpecusCore.LocalPeerService service;
    private final ServerSocket serverSocket;
    private final Set<InetAddress> allowedPeerAddresses;
    private final Set<String> auditedAccessEvents = ConcurrentHashMap.newKeySet();
    private final ExecutorService executor;
    private final AtomicBoolean open = new AtomicBoolean(true);
    private final ConcurrentHashMap<Socket, Socket> splices = new ConcurrentHashMap<>();
    private final Semaphore slots = new Semaphore(MAX_ACTIVE_CONNECTIONS);

    private PeerServiceBridge(String virtualIp, SpecusCore.LocalPeerService service, ServerSocket serverSocket) {
        this.virtualIp = virtualIp;
        this.service = service;
        this.serverSocket = serverSocket;
        this.allowedPeerAddresses = allowedPeerAddresses(service);
        this.executor = Executors.newCachedThreadPool(r -> {
            Thread thread = new Thread(r, "peer-service-bridge-" + service.serviceId);
            thread.setDaemon(true);
            return thread;
        });
        executor.execute(this::acceptLoop);
    }

    static PeerServiceBridge bind(String virtualIp, SpecusCore.LocalPeerService service) throws IOException {
        if (!PeerServiceRuntime.isLocalServiceTarget(service.targetHost)) {
            throw new IOException("targetHost is not assigned to this device");
        }
        ServerSocket server = new ServerSocket();
        server.setReuseAddress(true);
        server.bind(new InetSocketAddress(InetAddress.getByName(virtualIp), service.publishedPort));
        return new PeerServiceBridge(virtualIp, service, server);
    }

    boolean matches(String virtualIp, SpecusCore.LocalPeerService service) {
        return Objects.equals(this.virtualIp, virtualIp)
                && Objects.equals(this.service.serviceId, service.serviceId)
                && this.service.publishedPort == service.publishedPort
                && Objects.equals(this.service.targetHost, service.targetHost)
                && this.service.targetPort == service.targetPort
                && allowedPeerAddresses.equals(allowedPeerAddresses(service));
    }

    private void acceptLoop() {
        while (open.get() && !serverSocket.isClosed()) {
            try {
                Socket inbound = serverSocket.accept();
                if (!allowedPeerAddresses.contains(inbound.getInetAddress())) {
                    auditAccessOnce("deny", inbound.getRemoteSocketAddress(), "source-not-allowed");
                    closeQuietly(inbound);
                    continue;
                }
                auditAccessOnce("allow", inbound.getRemoteSocketAddress(), "acl-authorized");
                if (!slots.tryAcquire()) {
                    closeQuietly(inbound);
                    continue;
                }
                try {
                    executor.execute(() -> splice(inbound));
                } catch (RuntimeException rejected) {
                    slots.release();
                    closeQuietly(inbound);
                }
            } catch (IOException e) {
                return;
            }
        }
    }

    static Set<InetAddress> allowedPeerAddresses(SpecusCore.LocalPeerService service) {
        Set<InetAddress> addresses = new HashSet<>();
        if (service == null || service.allowedPeerVirtualIps == null) {
            return Set.of();
        }
        for (String raw : service.allowedPeerVirtualIps) {
            try {
                if (raw != null && !raw.isBlank()) {
                    addresses.add(InetAddress.getByName(raw.trim()));
                }
            } catch (Exception ignored) {
                // Invalid server-authored entries are ignored; empty is fail-closed.
            }
        }
        return Set.copyOf(addresses);
    }

    private void auditAccessOnce(String action, Object source, String reason) {
        String sourceAddress = String.valueOf(source);
        String key = action + '|' + sourceAddress;
        if (auditedAccessEvents.size() < 128 && auditedAccessEvents.add(key)) {
            LOG.info("[peer-service-access-audit] action=" + action
                    + " serviceId=" + service.serviceId
                    + " source=" + sourceAddress
                    + " reason=" + reason);
        }
    }

    private void splice(Socket inbound) {
        Socket outbound = new Socket();
        splices.put(inbound, outbound);
        try {
            inbound.setTcpNoDelay(true);
            outbound.setTcpNoDelay(true);
            outbound.connect(new InetSocketAddress(service.targetHost, service.targetPort), 3_000);
            Thread reply = new Thread(() -> copy(outbound, inbound), "peer-service-up-" + service.serviceId);
            reply.setDaemon(true);
            reply.start();
            copy(inbound, outbound);
            reply.join(1_000);
        } catch (Exception ignored) {
        } finally {
            closeQuietly(inbound);
            closeQuietly(outbound);
            splices.remove(inbound);
            slots.release();
        }
    }

    private static void copy(Socket from, Socket to) {
        try {
            InputStream in = from.getInputStream();
            OutputStream out = to.getOutputStream();
            byte[] buffer = new byte[8192];
            int read;
            while ((read = in.read(buffer)) >= 0) {
                out.write(buffer, 0, read);
                out.flush();
            }
        } catch (IOException ignored) {
            closeQuietly(from);
            closeQuietly(to);
        }
    }

    private static void closeQuietly(Socket socket) {
        try {
            socket.close();
        } catch (IOException ignored) {
        }
    }

    @Override
    public void close() {
        if (!open.compareAndSet(true, false)) {
            return;
        }
        LOG.info("[peer-service-access-audit] action=revoke serviceId=" + service.serviceId
                + " activeConnections=" + splices.size()
                + " reason=config-withdrawn-or-shutdown");
        try {
            serverSocket.close();
        } catch (IOException ignored) {
        }
        splices.forEach((inbound, outbound) -> {
            closeQuietly(inbound);
            closeQuietly(outbound);
        });
        splices.clear();
        executor.shutdownNow();
    }
}

File: implementations/csharp/client/src/Specus.Client/PeerMesh/PeerServiceBridge.cs
using System.Net;
using System.Net.Sockets;
using System.Collections.Concurrent;
using Microsoft.Extensions.Logging;
using Specus.Client.Configuration;

namespace Specus.Client.PeerMesh;

internal interface IPeerServiceForwarder : IDisposable
{
    string ServiceId { get; }
    bool Matches(string virtualIp, LocalPeerService service);
    (long BytesIn, long BytesOut, int Active, long Total) Snapshot();
}

internal sealed class PeerServiceBridge : IPeerServiceForwarder
{
    private readonly string _virtualIp;
    private readonly LocalPeerService _service;
    private readonly TcpListener _listener;
    private readonly CancellationTokenSource _cts = new();
    private readonly ILogger? _logger;
    private readonly ConcurrentDictionary<TcpClient, TcpClient> _splices = new();
    private readonly SemaphoreSlim _slots = new(64, 64);
    private readonly ConcurrentDictionary<string, byte> _auditedAccessEvents = new(StringComparer.Ordinal);
    private long _bytesIn;
    private long _bytesOut;
    private long _totalConnections;
    private int _disposed;

    private PeerServiceBridge(string virtualIp, LocalPeerService service, TcpListener listener, ILogger? logger)
    {
        _virtualIp = virtualIp;
        _service = service;
        _listener = listener;
        _logger = logger;
        _ = AcceptLoopAsync();
    }

    public static PeerServiceBridge Bind(string virtualIp, LocalPeerService service, ILogger? logger)
    {
        if (!PeerServiceDiscovery.IsLocalInterfaceTarget(service.TargetHost))
        {
            throw new InvalidOperationException("targetHost is not assigned to this device");
        }
        var listener = new TcpListener(IPAddress.Parse(virtualIp), service.PublishedPort);
        listener.Server.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, true);
        listener.Start();
        return new PeerServiceBridge(virtualIp, service, listener, logger);
    }

    public string ServiceId => _service.ServiceId;

    public (long BytesIn, long BytesOut, int Active, long Total) Snapshot() =>
        (Interlocked.Read(ref _bytesIn), Interlocked.Read(ref _bytesOut), _splices.Count,
            Interlocked.Read(ref _totalConnections));

    public bool Matches(string virtualIp, LocalPeerService service) =>
        _virtualIp == virtualIp
        && _service.ServiceId == service.ServiceId
        && _service.PublishedPort == service.PublishedPort
        && _service.TargetHost == service.TargetHost
        && _service.TargetPort == service.TargetPort
        && PeerServiceDiscovery.SameAllowedPeers(_service, service);

    private async Task AcceptLoopAsync()
    {
        while (!_cts.IsCancellationRequested)
        {
            TcpClient inbound;
            try
            {
                inbound = await _listener.AcceptTcpClientAsync(_cts.Token).ConfigureAwait(false);
            }
            catch (OperationCanceledException)
            {
                return;
            }
            catch
            {
                return;
            }
            if (inbound.Client.RemoteEndPoint is not IPEndPoint remote
                || !PeerServiceDiscovery.IsSourceAllowed(remote.Address, _service))
            {
                AuditAccessOnce("deny", inbound.Client.RemoteEndPoint, "source-not-allowed");
                inbound.Dispose();
                continue;
            }
            AuditAccessOnce("allow", remote, "acl-authorized");
            if (!_slots.Wait(0))
            {
                inbound.Dispose();
                continue;
            }
            _ = SpliceAsync(inbound);
        }
    }

    private void AuditAccessOnce(string action, EndPoint? source, string reason)
    {
        var sourceAddress = source?.ToString() ?? "unknown";
        var key = $"{action}|{sourceAddress}";
        if (_auditedAccessEvents.Count < 128 && _auditedAccessEvents.TryAdd(key, 0))
        {
            _logger?.LogInformation(
                "[peer-service-access-audit] action={Action} serviceId={ServiceId} source={Source} reason={Reason}",
                action, _service.ServiceId, sourceAddress, reason);
        }
    }

    private async Task SpliceAsync(TcpClient inbound)
    {
        using var inboundClient = inbound;
        using var outbound = new TcpClient();
        if (_cts.IsCancellationRequested || !_splices.TryAdd(inboundClient, outbound))
        {
            return;
        }
        Interlocked.Increment(ref _totalConnections);
        try
        {
            inboundClient.NoDelay = true;
            outbound.NoDelay = true;
            using var connectCts = new CancellationTokenSource(TimeSpan.FromSeconds(3));
            await outbound.ConnectAsync(_service.TargetHost, _service.TargetPort, connectCts.Token).ConfigureAwait(false);
            await using var inboundStream = inboundClient.GetStream();
            await using var outboundStream = outbound.GetStream();
            var up = CopyCountedAsync(inboundStream, outboundStream, incrementIn: true);
            var down = CopyCountedAsync(outboundStream, inboundStream, incrementIn: false);
            await Task.WhenAny(up, down).ConfigureAwait(false);
        }
        catch (Exception ex) when (ex is not OperationCanceledException)
        {
            _logger?.LogDebug(ex, "Peer-only 桥接转发失败 service={ServiceId}", _service.ServiceId);
        }
        finally
        {
            _splices.TryRemove(inboundClient, out _);
            _slots.Release();
        }
    }

    private async Task CopyCountedAsync(NetworkStream from, NetworkStream to, bool incrementIn)
    {
        var buffer = new byte[8192];
        while (!_cts.IsCancellationRequested)
        {
            var read = await from.ReadAsync(buffer, _cts.Token).ConfigureAwait(false);
            if (read <= 0)
            {
                return;
            }
            await to.WriteAsync(buffer.AsMemory(0, read), _cts.Token).ConfigureAwait(false);
            if (incrementIn)
            {
                Interlocked.Add(ref _bytesIn, read);
            }
            else
            {
                Interlocked.Add(ref _bytesOut, read);
            }
        }
    }

    public void Dispose()
    {
        if (Interlocked.Exchange(ref _disposed, 1) != 0)
        {
            return;
        }
        _logger?.LogInformation(
            "[peer-service-access-audit] action=revoke serviceId={ServiceId} activeConnections={ActiveConnections} reason=config-withdrawn-or-shutdown",
            _service.ServiceId, _splices.Count);
        _cts.Cancel();
        try
        {
            _listener.Stop();
        }
        catch
        {
            // ignored
        }
        foreach (var splice in _splices)
        {
            splice.Key.Dispose();
            splice.Value.Dispose();
        }
        _splices.Clear();
        _cts.Dispose();
    }
}
